hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java

```diff
@@ -493,15 +451,32 @@ public OzoneManagerDoubleBuffer buildDoubleBufferForRatis() {
    */
   @Override
   public long takeSnapshot() throws IOException {
-    LOG.info("Current Snapshot Index {}", getLastAppliedTermIndex());
-    TermIndex lastTermIndex = getLastAppliedTermIndex();
-    final TransactionInfo build = TransactionInfo.valueOf(lastTermIndex);
-    ozoneManager.setTransactionInfo(build);
-    Table<String, TransactionInfo> txnInfoTable =
-        ozoneManager.getMetadataManager().getTransactionInfoTable();
-    txnInfoTable.put(TRANSACTION_INFO_KEY, build);
+    // wait until applied == skipped
+    while (getLastAppliedTermIndex().getIndex() < lastSkippedIndex) {
+      try {
+        ozoneManagerDoubleBuffer.awaitFlush();
+      } catch (InterruptedException e) {
+        throw IOUtils.toInterruptedIOException("Interrupted ozoneManagerDoubleBuffer.awaitFlush", e);
+      }
+    }
+
+    return takeSnapshotImpl();
+  }
+
+  private synchronized long takeSnapshotImpl() throws IOException {
+    final TermIndex applied = getLastAppliedTermIndex();
+    final TermIndex notified = getLastNotifiedTermIndex();
+    final TermIndex snapshot = applied.compareTo(notified) > 0 ? applied : notified;
+    LOG.info(" applied = {}", applied);
+    LOG.info(" skipped = {}", lastSkippedIndex);
+    LOG.info("notified = {}", notified);
+    LOG.info("snapshot = {}", snapshot);
+
+    final TransactionInfo transactionInfo = TransactionInfo.valueOf(snapshot);
+    ozoneManager.setTransactionInfo(transactionInfo);
+    ozoneManager.getMetadataManager().getTransactionInfoTable().put(TRANSACTION_INFO_KEY, transactionInfo);
```

Comment:
updating TransactionInfoTable may have conflict if double buffer is also updating parallel same table. This can happen if ratis call applyTransaction parallel to takeSnapshot. Please check if this scenario is possible from ratis perspective.

Reply:
It won't happen in Ozone -- `takeSnapshotImpl() ` is `synchronized`.

Reply:
As discussed, before call snapshot,
1. applyTransaction executed previously is finished for response
2. double buffer is flushed
So based on this, no task pending which can cause parallel update